torch/optim/sparse_adam.py

```diff
@@ -66,8 +66,8 @@ def step(self, closure=None):
                 state['step'] += 1
 
                 grad = grad.coalesce()  # the update is non-linear so indices must be unique
-                grad_indices = grad._indices()
-                grad_values = grad._values()
+                grad_indices = torch._indices(grad)
```

Comment:
I think we should leave _indices and _values alone; sparse users use these fields a lot but we haven't changed the name from _* to something that doesn't have an _.

Reply:
Can we give them non-underscore names and make this a breaking change? This is certainly something we should fix and we do have a release upcoming, which gives us an opportunity.

What do you think? Also @soumith @SsnL

Reply:
cc @weiyangfb and @li-roy  for sparse things.

I feel uncomfortable breaking user code in one release; we should deprecate with warning and then break their code in a future release. On the other hand, idk how many people actually use sparse and we do promise no guarantees on our sparse API so...

Reply:
I am fine with non-underscore names and give deprecate warning at the current underscored APIs

Reply:
I'd argue that we should keep them.  For two reasons:

1. `_indices` and `_values` are really the properties/attributes of sparse tensors, so `torch._indices(tensor)` makes much less sense than `tensor._indices()`, at least to me. 

2. BC reasons as @zou3519 mentioned. Given how few sparse operations we support, many users of sparse tensors probably implemented their own sparse operations with `_indices` and `_values`.

FWIW, I am preparing a patch that adds `get_values` and `get_indices` methods, and those should be the public APIs, but they have subtle differences with `_values` and `_indices`, which make them much less useful in implementing custom sparse operations.

Reply:
What is we put them as methods on `Tensor` from within `torch/tensor.py`? As in, we'd add:

```
def _indices(self):
    return torch._indices(self)
```
That would preserve current users but keep the actual Tensor API clean from C++ side. I would still encourage you to come up with a deprecation strategy because calling an underscore method sounds/looks wrong either way, but that's up to you guys then. If there's any strong reason for keeping the method in C++ too we can make an exception for `_indices` and `_values`, but it sounds like a Python-only solution would work here?

Reply:
Many of the sparse methods implemented in aten use and mutate tensor._indices() / tensor._values(). It's not very nice looking at at::_indices(tensor) but I suppose as devs we can eat the notation.

Reply:
Ok why don't we (1) remove _indices/_values from Python (2) add `indices`/`values` (no underscore) (3) use `tensor.indices` in C++ (4) deprecate `_indices` in Python in favor of `indices`? Or some selection of these 4?

Reply:
I agree with @SsnL and @zou3519 , `_indices` and `_values` are fields of sparse tensor, so it is unintuitive to call `at::_indices(tensor)`. I am also fine with `tensor.indices()` and `tensor.values()` and deprecate `_*` methods. btw, @SsnL are `get_indices` and `get_values` going to do same thing as `_indices` and `_values`?

Reply:
Even in Python people can see/use these methods prefixed by underscores, and it hasn't caused us troubles. Why is this a problem for c++ API?

Reply:
It's not about trouble, it's just that we'd like the documentation for [at::Tensor](https://pytorch.org/cppdocs/api/structat_1_1_tensor.html#exhale-struct-structat-1-1-tensor) which is generated from the build/aten/src/ATen/Tensor.h file to be the exact public API we want users to use. In our Python docs we also filter out all the underscore methods except those three sparse "properties" https://imgur.com/a/Si2g5W5. I can make exceptions for those three, but just thinking about it, we should come up with a better API than an underscore method. It's not Pythonic at all!

Reply:
My main concern is not about doc/public API, but that
1. They are really really attributes of sparse tensors! So it makes very little sense to say `at::indices(tensor)` (or`_indices`).
2. They are really really useful in writing sparse ops! 
So I think one reasonable solution is to provide some new better-named way to access these fields, so our c++ devs/users can use thoseto write the kernels. 

Fortunately this is exactly what you new commit does! :D although I would name them as `get_raw_values` and `get_raw_indices`, to indicate that they are unsafe and should usually be avoided.

cc @ezyang if he has opinions on any of these.

Reply:
I believe the original reason why indices and values were prefixed with an underscore was because if you are working directly with `_values`, you have to be very careful about whether or not the sparse tensor is coalesced or not. Additionally, you can't differentiate through them, because `_indices` is defined to alias with the original tensor, but you can't do that if you have to wrap it in a new Variable to track the autograd (you also have differentiability problems if you allow returning duplicate entries!). @SsnL's pending patch closes these loopholes and gives you something that is semantically sound, but is less efficient than you would otherwise expect.